Specification Draft

stairman06 opened the pull request:
This is a specification draft based on previous discussions. Any feedback or thoughts are appreciated.

stairman06 committed:
First Spec Draft

immibis commented on format_introduction.md:
```diff
@@ -0,0 +1,23 @@
+
+## Design goals of the format
+
+The MCIP Format is designed to be:
```
What does it actually do? Who uses it and how?

immibis commented on format_introduction.md:
```diff
@@ -0,0 +1,23 @@
+# Introduction to the MCIP Format
+
+The MCIP Format is an open format for Minecraft assets. A Minecraft asset is a Mod, Resource/Texture Pack, World, or Modpack.
```
Asset -> project?

immibis commented on format_example.json:
```diff
@@ -0,0 +1,104 @@
+  "id": "dfa65597-eb1b-4127-9200-e53cc41d2d93",
+  "name": "Example Project",
+  "summary": "This is a short, one-sentence summary of the Example Project.",
+  "icon": "https://example.com/example.png",
```
Maybe move to gallery with rel: icon
This would reduce special cases

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+### `icon` (optional)
+
+A URL to an icon of the Project. It is RECOMMENDED to use JPEG or PNG images. Images encoded into the field, such as a Base64 encoded PNG, SHOULD NOT be used.
```
If one is used, how is it specified? Data URL?

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+### `description` (optional)
+
+A long description of the project. HTML and Markdown MAY be used. Any form of code execution such as JavaScript or WebAssembly SHOULD NOT be used. HTML forms, buttons, or other form of interactive elements SHOULD NOT be used. iframes to external domains MAY be used, but it is recommended to embed iframes only with a trustworthy website.
```
Markdown is bad for consistency. It will render incorrectly on every site other than the one it came from. How about XHTML?

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+##### `src`
+
+The `src` field is valid for both `video` and `image`. It contains a direct URI to either a playable video file or an image file. It is RECOMMENDED for images to be in PNG and JPEG formats, and it is RECOMMENDED for videos to be in the MP4 format.
```
Only proprietary video formats recommended?

stairman06 replied:
Was unaware that MP4 was proprietary, I'll replace it with WebM.

immibis replied:
Maybe there should also be a recommended codec, not just a container format. I don't know which codecs are typically used with WebM. Theora used to be the standard free software video codec, don't know if It still is.

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+### `license` (optional)
+
+This is an object that contains information about the license that a project is under. It contains the following fields:
```
Please consider adding SPDX identifier.

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+### `successor` and `predecessor` (optional)
+
+If a newer version of a Project has been released and it does not have the same ID as the previous version, the `successor` and `predecessor` fields may be set for the appropriate Projects. This fields MUST contain the ID of the predecessor or successor project.
```
Can there be more than one?

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+## Version Information
+
+Projects MAY contain a `versions` array, which is made up of Version Objects. If you are writing an API that uses the MCIP format, this field MAY be separated into a separate API call in order to save bandwidth. The `versions` field MAY also be paginated in API calls.
```
This brings up the question of what the specification actually describes. Presumably the API can separate out whatever its designers want it to. For example my API would probably separate out all the non-machine-readable content like descriptions and galleries from the machine-readable content.

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+### `semver` (optional)
+
+An optional Semantic Versioning compatible version number.
```
Does it have to be related to the actual mod version or is this just for better tracking compatibility? Can it change later if an incompatibility is found?

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+- `modsDirectory` - This file must be placed in the `mods` directory
+- `jarMod` - This file is a Jar mod and must be added into minecraft.jar
+- `other`
```
What does it mean if this is not present?
What about coremods directory (old versions of forge)?

Mods might even have multiple parts. I don't know any mod that did this, but perhaps a mod would have a coremod file and a normal mod file as well.

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+# API and Field Modularization Advice
+
+The MCIP format includes many fields. It would be innefficient for Hosts and API providers to return all information about a Project at once. Because of this, it's recommended to follow these recommended guidelines:
```
Typo: inefficient

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+---
+
+# API and Field Modularization Advice
```
I don't think this section is needed, if we can clearly define what this spec is for

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+##### `sha256` (optional)
+
+An SHA256 checksum of the file. This may be used by launchers in order to verify download integiry.
```
I recommend making it mandatory. Otherwise, a link in the index WILL break and everyone will be downloading the error page and saving it to the mods folder.

It's less critical for media, but I also suggest a checksum for gallery files (not embeds).

Another advantage: if the link breaks, we can still find the right file in the archives based on its checksum. We don't have to track which files came from which links at which time.

immibis commented on format_example.json:
```diff
@@ -0,0 +1,104 @@
+  "license": {
+    "name": "MIT",
+    "uri": "https://opensource.org/licenses/MIT",
+    "modpackLicense": "alwaysAllow"
```
Spec calls it modpackUsage

immibis commented on format_example.json:
```diff
@@ -0,0 +1,104 @@
+  "summary": "This is a short, one-sentence summary of the Example Project.",
+  "icon": "https://example.com/example.png",
+  "description": "This is a long description describing the project. <b>HTML</b> and **Markdown** can optionally be used.",
+  "gallery": [
```
I'd rename to "media". Like how the links section is called links, not sidebar.

immibis commented on format_example.json:
```diff
@@ -0,0 +1,104 @@
+    {
+      "name": "MCIP Contributors",
+      "uri": "https://github.com/mc-cip"
+    }
```
Shall we have a UUID to allow searching by author?

Shall we allow multiple UUIDs so that "cofh team" can link to all their individual profiles as well? Or is that too complicated?

If authors are added or removed is there a way to see which authors wrote a particular version?

immibis commented on format_example.json:
```diff
@@ -0,0 +1,104 @@
+        {
+          "type": "framework",
+          "id": "fabric-loader",
+          "version": ">= 0.6.1.45"
```
If you have fabric and forge versions, do they both count as different files in the same version, or different versions? If it's different versions then files/frameworks is useless. If it's different files then the dependency information will be wrong.

Also what makes frameworks different from projects?

immibis commented on format_example.json:
```diff
@@ -0,0 +1,104 @@
+    {
+      "name": "Example Category",
+      "uri": "https://example.com/browse?category=example",
+      "icon": "https://example.com/example-category.png"
```
This is a bit pedantic, but the category icon could have a checksum.

Minenash replied:
I'm not sure if category information should even be a mod spec file. Just a list of categories should be all that's needed from a mod's spec

stairman06 replied:
It probably is redundant to include category icons and URIs in each mod - I may replace it with a list of category names, and instead have a map between category names and icons/URIs returned separately.

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+##### `download`
+
+This field MUST be an array listing available download sources for this file. Each object containing in this array MUST contain the `uri` field, which is a URL pointing to a direct download of the file.
+
```
Are source code and deobf files also supported? How are they marked? Rel for files perhaps.

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,289 @@
+
+##### `name`
+
+The full name of this file, including file extension.
```
Windows and Unix special characters are prohibited. Unicode allowed. Windows device filenames are prohibited. Filename must not begin or end with a dot (this includes . and ..) or whitespace or be blank.

Or, we can say that anything is allowed here but the installer doesn't have to honour it.

stairman06 committed:
Refinements based on feedback

stairman06 commented:
Made some modifications based on feedback:
- Icons are now in the `media` array
- Media objects can now have an SHA256 hash
- `authors` supports Teams
- Authors have IDs
- Projects can have multiple successors and predecessors
- WebM is now recommended over MP4
- Dependencies can now be both for the Version and for the File
- The `file.frameworks` field has been removed
- Files now have a `rel` field
- `installationMethod` now has support for any directory
- `download` field has been simplified and renamed to `downloads`

immibis commented on format_example.json:
```diff
@@ -0,0 +1,135 @@
+    {
+      "rel": "icon",
+      "uri": "https://example.com/example.png",
+      "sha256": "< sum goes here >"
```
Is type also needed?
Also this has a uri while other files have a src

immibis commented on format_example.json:
```diff
@@ -0,0 +1,135 @@
+    {
+      "team": true,
+      "id": "111111-111111-111111-111111-111111",
+      "name": "MCIP Contributors",
```
Does the team also have a URI? I assume thile intention is to open this URI when you click on the author's name

immibis commented on format_example.json:
```diff
@@ -0,0 +1,135 @@
+    }
+  ],
+  "license": {
+    "spdx": "MIT",
```
Only common open source licenses have SPDX identifiers, so we will also need to:

* Define an extension to our version of SPDX which lists more licenses, including "unrecognised open licence" and "unrecognised proprietary license" as well as things like MMPL
* Add a name as well as SPDX. I think SPDX should be mandatory, since you can always use an unrecognised license identifier if we define one. I don't think SPDX defines one.

stairman06 replied:
Added MMPL as well as Generic-Open and Generic-Proprietary licenses - are there any more that should be added?

immibis commented on format_example.json:
```diff
@@ -0,0 +1,135 @@
+            }
+          ],
+          "installationMethod": "placeInDirectory",
+          "directory": "mods",
```
Should go inside installationMethod

RyanTheAllmighty replied:
~Yup, lower down also lists `"installationMethod": "modsDirectory",` so are they interchangeable?~

~Consistency would be good.~

Answered pretty much at the bottom.

Installation instructions can be quite complex, depending on the mod and file itself, so having `installationMethod` as an object with deeper information might be valuable

stairman06 replied:
I'll change `installationMethod` to an object and remove `modsDirectory`.

What other information could be added? Mentioned below was frameworks like Forge being provided as a project, so how could this be represented in `installationMethod`? Framework installation can be complicated, especially 1.13+ Forge.

immibis replied:
I suppose for Forge the installation method is forgeInstaller and for Fabric it's fabricInstaller. But this is not true for all versions! Old versions of Forge were just jar mods.

stairman06 replied:
forgeInstaller and fabricInstaller will probably be the best options. Launchers may choose to ignore the value as many don't run installers, but it's good to have the metadata there anyway.

comp500 replied:
Fabric installation is fairly simple, all it really needs is a few extra libraries downloaded and a different mainClass, the installer just puts the vanilla launcher version json in the right place. The Fabric meta API has an endpoint for this https://meta.fabricmc.net/v2/versions/loader/1.16.1/0.9.0+build.204/profile/json. Forge installation since 1.13 pretty much requires an installer unless you use something like https://github.com/ZekerZhayard/ForgeWrapper (which just runs the installer on startup)

immibis replied:
Good point - should we include JSON library dependencies (I'll call them classpath dependencies)? Can any mod have a classpath dependency, or just frameworks? Minecraft launcher wants a very specific format for classpath dependencies but theoretically any mod could want one, especially jar mods.

stairman06 replied:
Classpath dependencies are interesting - currently there's the `versionJsonInstall` installation method which tells launchers to look at the file with `rel` set to `versionJson`, which works for Fabric but it uses the Minecraft launcher's format.

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+### `summary` (optional)
+
+A short, one sentence summary of the Project. Any form of rich text such as HTML SHOULD NOT be used.
```
If I get a file with HTML in this field, should I display it as text, display it as HTML, or remove the HTML?

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+##### `rel` (optional)
+
+The relation of this file. Currently, the only valid option is `icon`. If the Media Item is not the Project's icon, this field SHOULD NOT be present.
```
I think the same rules should apply for all rels in the format: you can have more than one in an array; empty array is the same as not present. (Even though only one is known now, more could be added later. What about rel=gallery?)

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+##### `embed`
+
+The `embed` field is valid for both `video` and `image`. It contains a URI to a webpage which SHOULD be embedded when displaying this Media Item. It is recommended for hosts to only allow trustworthy domains to be embedded.
```
Is there any difference between embedding a video and an image?

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+This is an object that contains the following fields:
+
+##### `sha256` (optional)
```
Should be a new heading level under icon, unless markdown only supports 3 levels

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+### `semver` (optional)
+
+An optional Semantic Versioning compatible version number. This field SHOULD ONLY be present if the version's name conforms to Semantic Versioning. This field SHOULD NOT be present if the name does not conform to SemVer.
```
So it can still be present if the name doesn't conform to semver? Why SHOULDN'T it? Does it have to be equal to the name?

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+## Version Information
+
+Projects can contain a `versions` array, which lists Version Objects.
```
Does the order matter? Especially since some projects use version numbers that aren't machine-comparable and some may reset version numbers or change numbering schemes. Newest to oldest or oldest to newest?

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+### `releaseDate`
+
+The release date of this version. It MUST be stored as an ISO-8601 conforming String.
```
Does it include time? Timezonr?

comp500 replied:
ISO8601 can include time and timezone, so I'd think an ISO-8601 conforming String would include that.

stairman06 replied:
I was thinking to use UTC (e.g. `2020-08-05T12:00:00Z` is 12:00 UTC).

waitwhereami replied:
I've done a non-trivial amount of work with dates and timezones and the key thing to determine how to store the data amounts to "is this a thing that happened/is happening. or a thing that will happen?"

If it is something that happened in the past, or a record of something happening (timestamp, event log) then UTC is probably the easiest and safest bet. If it is happening in the future, you need to log the actual desired timezone along with the ISO-8601 data. ISO-8601 actually does not attempt to store timezone data. There is an optional "Z" indicator which is supposed to indicate that the time is in UTC; in practicality, nearly all ISO-8601 datetimes are in UTC so most systems ignore the Z.

Some systems store the offset of the timezone along with the date, but that is unfortunately also insufficient. If someone wants to record an event occurring on 2024/06/04T08:30:00+05:00, but the local government decides to stop observing DST between storing the date and retrieving it, then an inconsistency is introduced.

immibis replied:
Since we are only recording things that happened, then, it's UTC.

The documentation should give an example value. xxxx-xx-xxTxx:xx:xxZ. No others. According to Wikipedia, 2020-08-05 is also an ISO-8601 conforming string. So is 2020-W32. So is --08-05. So is 2020-218. But we only want a particular format.

If the information is not known, should it be approximated? How much accuracy is needed?

stairman06 replied:
I'm working on modifying the documentation and providing an example value, `2020-01-01T12:00:00Z`.

If information about the time a version was released is unknown, it should default to 12:00:00 UTC. I don't know how much precision should go with the actual date, but at some point the value can default to `unknown`. The defaulting of release dates should fall to hosts, and launchers shouldn't have to generate default date values.

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+### `changelog`
+
+The changelog of this version. This MAY be moved into a separate API call for hosts. The changelog MAY contain HTML.
```
Literally everything MAY be a separate API call so I still don't think this is relevant here.

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+- `asset` - For example, a required library mod
+- `framework` - For example, a required modloader
+
+Frameworks exist as a separate type because they may require a more advanced installation process.
```
Isn't that a problem for the framework's project entry to deal with, not the project using the framework?

stairman06 replied:
Frameworks are separate because they may not have a project entry. I personally don't think Forge will provide MCIP format-compatible data any time soon.

immibis replied:
So the Minecraft Index Project isn't going to include Minecraft Forge in the index? ...because?

stairman06 replied:
I thought about this more and it may be better to store them as projects. I'll add some more fields and values to support it.

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+- `homepage` - The official homepage of a Project
+- `documentation` - Documentation of a Project.
+- `wiki` - Wiki information about a Project.
```
Note that wiki is just a less generic name for documentation in most cases. I support using both generic and specific rel tags so that sites can change the link text for example. "Project Wiki" instead of "Project Docs" or "Project GitHub" instead of "Source code"

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+- `fabric-loader`
+- `forge`
+- `liteloader`
+- `rift`
```
Maybe this list goes in a separate document

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+The installation method of this file. Valid options are:
+
+- `placeInDirectory` - This file must be placed in the directory specified in a separate `directory` field
+- `modsDirectory` - A shorthand for `placeInDirectory` with `"directory": "mods"` as it's very common
```
Unnecessary extra work to implement. Just use placeInDirectory/mods

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+- `source` - Source code of this version
+- `deobf` - Deobfuscation file for this version
+
+If this field is not present, the file should be assumed to be `primary`. The values of this field CAN be present in multiple files. For example, a version can have a file for both the Forge and Fabric modloaders, and both of those files can have the `rel` field set to `primary`.
```
Why should that be assumed? Is it just in case the index team is too lazy to add it?

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+A field describing usage of the Project in modpacks. Valid options:
+
+- `alwaysAllow` - This Project can be used in a modpack without asking for permission.
```
Does it override the license? If modpackUsage is alwaysAllow but the license says no redistribution, is the mod allowed to be redistributed as part of a modpack?

immibis commented on format_spec.md:
```diff
@@ -0,0 +1,339 @@
+
+##### `src`
+
+This field contains where this dependency can be downloaded. It contains the name of the host, in an all-lowercase string. Examples include: `curseforge`, `diluv` and `modrinth`. If the `type` field is set to `framework`, this field CANNOT be present.
```
If the goal is to index all mods then the dependencies should already be indexed and src isn't needed. How is a launcher supposed to interpret src when it says, for example, `immibis.com` or `railcraft`?

stairman06 committed:
More revisions and examples based on feedback

stairman06 commented:
I've made some more revisions:
- The `src` value for files contains more information, and now mentions the API made by immibis (PR #3)
- `installationMethod` has been renamed to `installation` and is now an object.
- `installation` can now be at the version level and file level.
- `releaseDate` now contains more information about the ISO-8601 format and provides an example.
- An example for a framework using the MCIP Format, format_example_fabric, has been created.
- More license information and values have been added.
- The format_values.md file contains a list of different values that fields can have, such as standard framework names and file rel values.
- format_introduction now describes the standard API and provides information on how launchers can use this
- Media Items can now have mutliple `rel`s
- Dependencies can now have multiple `src` values
- `modpackUsage` now overrides any restrictions set in the license
- `installation` is now a required field - it MUST be present at either the version level or the file level
- More info about `semver` and the `versions` object

stairman06 committed:
Fix examples not adhering to spec

immibis commented:
You put `"icon": "image"` instead of `"type": "image"`

stairman06 committed:
Fix "icon": "Image"

immibis commented:
Can we move the installation data to a different PR since it's a lot more complex than we thought?

stairman06 commented:
Created PR #7 for discussion of installation data.

stairman06 committed:
Add reference to installation data PR

stairman06 committed:
Classpath library + Replace minecraftVersions

stairman06 committed:
HTML implementation guide

stairman06 committed:
Categories -> Tags, cleanup dependencies

stairman06 commented:
Did some more work:
- Categories become Tags and there are now some standardized values
- Moved guides/tips for launchers to a separate file and added HTML guide
- Dependency cleanup
- Removal of references to MCIP Standardized API as it had little purpose
     - Redefined dependency `src` field to allow offsite dependency metadata
- Grammatical fixes
- General cleanup and fixing
- Add `compressedOverrides` file `rel` value

stairman06 committed:
Add JSON Schema

stairman06 committed:
Move examples to subdirectory

stairman06 committed:
Example modpack

stairman06 committed:
Conditions! For clientside and serverside info

stairman06 commented:
Now added: conditions. Conditions allow you to describe where a Project can be used. Currently, the only available condition options are clientside and serverside. Conditions can either be defined in a separate `conditions` field, or directly in a field. You can find examples in format_spec.md, line 278. An example is also in format_example_modpack.json

stairman06 committed:
Remove "optional" value for "allowed"

stairman06 committed:
Slight condition improvements

stairman06 committed:
MCIP -> MODIP

stairman06 commented:
Repository abandoned. Moved to [modip-org/spec](https://github.com/modip-org/spec).
